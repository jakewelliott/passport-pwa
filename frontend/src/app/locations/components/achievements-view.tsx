import type { Park, ParkActivity } from '@/lib/mock/types';
import { FaRegCheckSquare, FaRegSquare, FaStamp } from 'react-icons/fa';

interface AchievementsViewProps {
  parkActivity: ParkActivity;
  park: Park;
}

const AchievementsView = ({ parkActivity, park }: AchievementsViewProps) => {
  return (
    <div className='flex flex-col gap-3' data-testid='achievements-view'>
      {parkActivity.stampCollectedAt !== undefined && (
        <div className='top-0 flex'>
          <FaStamp size={'17px'} strokeWidth={3} style={{ paddingRight: '5px', paddingTop: '5px' }} />
          <p>
            {parkActivity.stampCollectedAt
              ? `Stamp collected ${parkActivity.stampCollectedAt}`
              : 'Stamp not yet collected'}
          </p>
        </div>
      )}
      {
        park.bucketListItems &&
        park.bucketListItems.length > 0 && (
          <div className='top-0 flex'>
            {parkActivity.completedBucketListItems[0]?.id ? (
              <FaRegCheckSquare size={'17px'} strokeWidth={3} style={{ paddingRight: '5px', paddingTop: '5px' }} />
            ) : (
              <FaRegSquare size={'17px'} strokeWidth={3} style={{ paddingRight: '5px', paddingTop: '5px' }} />
            )}
            <p data-testid={'BLI'}>
              Bucket List Item:
              <br />
              {park.bucketListItems[0].task}
            </p>
          </div>
        )}
    </div>
  );
};

export default AchievementsView;
