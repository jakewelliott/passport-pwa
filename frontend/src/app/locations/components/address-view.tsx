import type { Address, Park } from '@/types';
import { useState } from 'react';
import { FiNavigation } from 'react-icons/fi';


const ParkAddress = ({ address }: { address: Address }) => {
	return (
		<div>
			<p>{address.title}</p>
			<p>{address.addressLineOne}</p>
			{address.addressLineTwo && <p>{address.addressLineTwo}</p>}
			<p>{address.city}, {address.state} {address.zipcode}</p>
		</div>
	);
};

export const AddressView = ({ park }: { park: Park }) => {

	const [showAll, setShowAll] = useState(false);
	const showButton = park.addresses && park.addresses.length > 1;

	const toggle = () => {
		setShowAll(!showAll);
	};

	const OtherAddresses = () =>
		park.addresses.slice(1).map((address) => (
			<ParkAddress key={address.title} address={address} />
		))


	const Button = () =>
		<button type='button' onClick={toggle} className='ml-4 inline-flex items-center'>
			{showAll ? 'Show Less' : 'Show More'}
		</button>

	return (
		<div className='top-0 flex' key={park.parkName}>
			<FiNavigation size={'17px'} strokeWidth={3} style={{ paddingRight: '5px', paddingTop: '5px' }} />
			<ParkAddress address={park.addresses[0]} />
			{showAll && <OtherAddresses />}
			{showButton && <Button />}
		</div>
	);
};
