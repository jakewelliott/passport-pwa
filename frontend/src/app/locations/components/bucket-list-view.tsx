import { useBucketList } from '@/hooks/queries/useBucketList';
import { useStamp } from '@/hooks/queries/useStamps';
import DateHelper from '@/lib/date-helper';
import type { BucketListItem, CollectedStamp, Park } from '@/lib/mock/types';
import { FaRegCheckSquare, FaRegSquare, FaStamp } from 'react-icons/fa';

const BucketListItemView = ({ bucketListItem }: { bucketListItem: BucketListItem }) => {
	const isCompleted = bucketListItem.completedAt && !bucketListItem.deleted;
	const Checkbox = isCompleted ? FaRegCheckSquare : FaRegSquare;

	return (
		<div>
			<Checkbox size={'17px'} strokeWidth={3} style={{ paddingRight: '5px', paddingTop: '5px' }} />
			<p data-testid={'BLI'}>
				Bucket List Item:
				<br />
				{bucketListItem.task}
			</p>
		</div>
	);
};

const StampView = ({ stamp }: { stamp: CollectedStamp | undefined }) => {
	return (
		<>
			<FaStamp size={'17px'} strokeWidth={3} style={{ paddingRight: '5px', paddingTop: '5px' }} />
			<p>
				{stamp
					? `Stamp collected on ${DateHelper.stringify(new Date(stamp.createdAt)).replace(',', ' at')}`
					: 'Stamp not yet collected'}
			</p>
		</>
	);
};

export const BucketListView = ({ park }: { park: Park }) => {
	const { data: items, isLoading: itemsLoading } = useBucketList(park.id);
	const { data: stamp, isLoading: stampLoading } = useStamp(park.abbreviation);

	// TODO: replace with a placeholder
	if (itemsLoading || stampLoading) return null;

	return (
		<div className='flex flex-col gap-3' data-testid='achievements-view'>
			<div className='top-0 flex'>
				<StampView stamp={stamp} />
			</div>
			<div className='top-0 flex'>
				{items.map((item: BucketListItem) => (
					<BucketListItemView key={item.id} bucketListItem={item} />
				))}
			</div>
		</div>
	);
};
