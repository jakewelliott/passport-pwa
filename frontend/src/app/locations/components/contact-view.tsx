import type { Park } from '@/types';
import { FaRegEnvelope } from 'react-icons/fa';
import { FiNavigation, FiPhone } from 'react-icons/fi';

export const ContactView = ({ park }: { park: Park }) => {
    return (
        <div className='mx-4 flex flex-col gap-3' data-testid='location-contact'>
            {/* Coordinates */}
            <div className='top-0 flex'>
                <FiNavigation size={'17px'} strokeWidth={3} style={{ paddingRight: '5px', paddingTop: '5px' }} />
                <p>
                    GPS: {park.coordinates.latitude.toFixed(4)}, {park.coordinates.longitude.toFixed(4)}
                </p>
            </div>
            {/* Phone */}
            <div className='top-0 flex'>
                <FiPhone size={'17px'} strokeWidth={3} style={{ paddingRight: '5px', paddingTop: '5px' }} />
                <p>{String(park.phone).replace(/(\d{3})(\d{3})(\d{4})/, '($1) $2-$3')}</p>
            </div>
            {/* Email */}
            <div className='top-0 flex'>
                <FaRegEnvelope size={'17px'} strokeWidth={3} style={{ paddingRight: '5px', paddingTop: '5px' }} />
                <p>{park.email}</p>
            </div>
        </div>
    );
};
