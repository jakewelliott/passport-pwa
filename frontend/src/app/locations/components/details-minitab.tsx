import type { Park } from '@/types';
import React from 'react';

export const DetailsMiniTab = ({ park }: { park: Park }) => {
	const renderTrails = (trails: string) => {
		return trails
			.split('\n')
			.filter((line) => line.trim() !== '') // Remove empty lines
			.map((line, index, array) => (
				<React.Fragment key={line}>
					{line}
					{index < array.length - 1 && <br />} {/* Add <br /> only if it's not the last line */}
				</React.Fragment>
			));
	};

	return (
		<div className='mt-6 mb-6 flex flex-col'>
			<p className='pr-6 pl-6'>
				{park.establishedYear && (
					<>
						<span className='green-text'>Established: </span>
						{park.establishedYear}
						<br />
						<br />
					</>
				)}
				{park.landmark && (
					<>
						<span className='green-text'>Landmark: </span>
						{park.landmark}
						<br />
						<br />
					</>
				)}
				{park.youCanFind && (
					<>
						<span className='green-text'>You can find... </span>
						{park.youCanFind}
						<br />
						<br />
					</>
				)}
				{park.trails && (
					<>
						<span className='green-text'>Trails: </span>
						<br />
						{renderTrails(park.trails)}
						<br />
						<br />
					</>
				)}
			</p>
			<div data-testid='icon-scroll-container' className='icon-scroll-container overflow-x-auto'>
				<div className='inline-flex gap-6 px-6'>
					{park.icons?.map((icon, index) => (
						<img
							src={`/icons/park/${icon.iconName}.svg`}
							width={55}
							height={55}
							key={icon.iconName}
							alt={`Park icon ${index + 1}`}
						/>
					))}
					<div className='w-px flex-shrink-0' />
				</div>
			</div>
		</div>
	);
};
