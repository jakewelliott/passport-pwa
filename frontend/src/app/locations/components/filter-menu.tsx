import { FaFilter } from 'react-icons/fa';

interface FilterMenuProps {
    searchQuery: string;
    setSearchQuery: (value: string) => void;
    setIsFilterModalOpen: (value: boolean) => void;
}

export const FilterMenu = ({ searchQuery, setSearchQuery, setIsFilterModalOpen }: FilterMenuProps) => (
    <div className='relative flex gap-2'>
        <div className='relative flex-1'>
            <input
                type='text'
                placeholder='Search parks...'
                value={searchQuery}
                onChange={(e) => setSearchQuery(e.target.value)}
                className='w-full rounded-lg border border-system-gray bg-supporting-lightblue p-3 pr-7 focus:border-secondary-darkteal focus:outline-none focus:ring-1 focus:ring-secondary-darkteal focus:ring-opacity-100'
            />
            {searchQuery && (
                <button
                    type='button'
                    onClick={() => setSearchQuery('')}
                    className='-translate-y-1/2 absolute top-1/2 right-3 transform bg-supporting-lightblue text-system-gray hover:text-secondary-darkteal'
                    aria-label='Clear search'
                >
                    &times;
                </button>
            )}
        </div>
        <button
            type='button'
            onClick={() => setIsFilterModalOpen(true)}
            className={
                'w-auto rounded-lg border border-system-gray bg-system-white px-3 py-2 transition-all duration-200'
            }
        >
            <FaFilter />
        </button>
    </div>
);
