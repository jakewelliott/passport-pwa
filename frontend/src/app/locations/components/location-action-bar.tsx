import type { Park } from '@/types';
import { FiGlobe, FiMail, FiNavigation, FiPhone } from 'react-icons/fi';

export const LocationActionBar = ({ park }: { park: Park }) => {
    // TODO: styling here needs to be fixed
    return (
        <div
            className='flex flex-row items-center justify-evenly bg-secondary-darkteal py-2.5'
            data-testid='location-action-bar'
        >
            {park.coordinates?.latitude && park.coordinates.longitude && (
                <a
                    href={`https://www.google.com/maps/place/${park.coordinates.latitude},${park.coordinates.longitude}`}
                    className='text-supporting-inactiveblue'
                >
                    <FiNavigation size={24} strokeWidth={3} />
                </a>
            )}
            {park.phone && (
                <a href={`tel://${park.phone}`} className='text-supporting-inactiveblue'>
                    <FiPhone size={24} strokeWidth={3} />
                </a>
            )}
            {park.website && (
                <a href={park.website} className='text-supporting-inactiveblue'>
                    <FiGlobe size={24} strokeWidth={3} />
                </a>
            )}
            {park.email && (
                <a href={`mailto:${park.email}`} className='text-supporting-inactiveblue'>
                    <FiMail size={24} strokeWidth={3} />
                </a>
            )}
        </div>
    );
};
