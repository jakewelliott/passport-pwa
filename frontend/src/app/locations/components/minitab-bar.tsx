import { useState } from 'react';
import { a11yOnClick } from '@/lib/a11y';

interface LocationMiniTabBarProps {
  children: React.ReactNode[];
}

// TODO: replace with a library (react-tabs)
export const LocationMiniTabBar: React.FC<LocationMiniTabBarProps> = ({ children }) => {
  const [choice, setChoice] = useState(0);

  return (
    <>
      <div className='flex h-12 items-center justify-around border-supporting_bluegray border-b'>
        {['Details', 'Photos', 'Notes'].map((tab, index) => (
          <div
            key={tab}
            {...a11yOnClick(() => setChoice(index))}
            className={`relative flex h-full w-full cursor-pointer items-center justify-center p-2 ${
              choice === index ? 'text-system_black' : 'text-supporting_darkgray'
            }`}
          >
            <p>{tab}</p>
            <div
              className={`absolute right-0 bottom-0 left-0 h-0.5 bg-secondary_darkteal ${choice !== index ? 'hidden' : ''}`}
            />
          </div>
        ))}
      </div>
      {children[choice]}
    </>
  );
};
