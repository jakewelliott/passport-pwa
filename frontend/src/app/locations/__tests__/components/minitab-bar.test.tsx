import { LocationMiniTabBar } from '@/app/locations/components/minitab-bar';
import { setupTestEnv } from '@/lib/testing/test-wrapper';
import { fireEvent, screen } from '@testing-library/react';
import { describe, expect, it } from 'vitest';

const mockChildren = [
    <div key='details'>Details Content</div>,
    <div key='photos'>Photos Content</div>,
    <div key='notes'>Notes Content</div>,
];

const { render } = setupTestEnv();

describe('LocationMiniTabBar', () => {
    it('matches snapshot', () => {
        const { container } = render(<LocationMiniTabBar>{mockChildren}</LocationMiniTabBar>);
        expect(container).toMatchSnapshot();
    });

    it('renders all tab buttons', () => {
        render(<LocationMiniTabBar>{mockChildren}</LocationMiniTabBar>);
        expect(screen.getByText('Details')).toBeTruthy();
        expect(screen.getByText('Photos')).toBeTruthy();
        expect(screen.getByText('Notes')).toBeTruthy();
    });

    it('shows first tab content by default', () => {
        render(<LocationMiniTabBar>{mockChildren}</LocationMiniTabBar>);
        expect(screen.getByText('Details Content')).toBeTruthy();
        expect(screen.queryByText('Photos Content')).not.toBeTruthy();
        expect(screen.queryByText('Notes Content')).not.toBeTruthy();
    });

    it('switches content when clicking different tabs', () => {
        render(<LocationMiniTabBar>{mockChildren}</LocationMiniTabBar>);

        // Click Photos tab
        fireEvent.click(screen.getByText('Photos'));
        expect(screen.queryByText('Details Content')).not.toBeTruthy();
        expect(screen.getByText('Photos Content')).toBeTruthy();
        expect(screen.queryByText('Notes Content')).not.toBeTruthy();

        // Click Notes tab
        fireEvent.click(screen.getByText('Notes'));
        expect(screen.queryByText('Details Content')).not.toBeTruthy();
        expect(screen.queryByText('Photos Content')).not.toBeTruthy();
        expect(screen.getByText('Notes Content')).toBeTruthy();
    });

    // it('applies active styles to selected tab', () => {
    //     render(<LocationMiniTabBar>{mockChildren}</LocationMiniTabBar>);

    //     const detailsTab = screen.getByText('Details').parentElement;
    //     const photosTab = screen.getByText('Photos').parentElement;

    //     expect(detailsTab).toHaveClass('text-system_black');
    //     expect(photosTab).toHaveClass('text-supporting_darkgray');

    //     fireEvent.click(screen.getByText('Photos'));

    //     expect(detailsTab).toHaveClass('text-supporting_darkgray');
    //     expect(photosTab).toHaveClass('text-system_black');
    // });

    // it('shows indicator line for active tab', () => {
    //     render(<LocationMiniTabBar>{mockChildren}</LocationMiniTabBar>);

    //     const tabs = screen.getAllByText(/Details|Photos|Notes/);
    //     const detailsIndicator = tabs[0].parentElement?.querySelector('.bg-secondary_darkteal');
    //     const photosIndicator = tabs[1].parentElement?.querySelector('.bg-secondary_darkteal');

    //     expect(detailsIndicator).not.toHaveClass('hidden');
    //     expect(photosIndicator).toHaveClass('hidden');

    //     fireEvent.click(screen.getByText('Photos'));

    //     expect(detailsIndicator).toHaveClass('hidden');
    //     expect(photosIndicator).not.toHaveClass('hidden');
    // });
});
