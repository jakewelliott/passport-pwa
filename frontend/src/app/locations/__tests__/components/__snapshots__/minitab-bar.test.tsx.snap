// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`LocationMiniTabBar > matches snapshot 1`] = `
<div>
  <div
    class="flex h-12 items-center justify-around border-supporting-bluegray border-b"
  >
    <div
      class="relative flex h-full w-full cursor-pointer items-center justify-center p-2 text-system-black"
    >
      <p>
        Details
      </p>
      <div
        class="absolute right-0 bottom-0 left-0 h-0.5 bg-secondary-darkteal "
      />
    </div>
    <div
      class="relative flex h-full w-full cursor-pointer items-center justify-center p-2 text-supporting-darkgray"
    >
      <p>
        Photos
      </p>
      <div
        class="absolute right-0 bottom-0 left-0 h-0.5 bg-secondary-darkteal hidden"
      />
    </div>
    <div
      class="relative flex h-full w-full cursor-pointer items-center justify-center p-2 text-supporting-darkgray"
    >
      <p>
        Notes
      </p>
      <div
        class="absolute right-0 bottom-0 left-0 h-0.5 bg-secondary-darkteal hidden"
      />
    </div>
  </div>
  <div>
    Details Content
  </div>
</div>
`;
